### PR TITLE
Memory allocation allocation tracker test helper

### PR DESCRIPTION
**Pull request by tocsoft**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

This adds some new test helpers APIs + updates the `MemoryDiagnostics` class to consistently track the memory buffers per test.

Also adds 2 helpers in to the tests 
1. `ValidateDisposedMemoryAllocationsAttribute` - This attributes a can be applied to test classes and `Fact`/`Theory`s directly to automatically apply the memory application tracking and validation.
2. `MemoryAllocatorValidator` static class can be used in places the attribute is not appropriate, i.e. inside Remote Executors.

I've also applied some fixes targeting the Decoder tests

this includes the fix from #2081 (which I will merge into this PR so we don't loose attribute of the fix) if we decide this approach is suitable...otherwise we should just merge that one.

### CODE REVIEW

**Comment by tocsoft** [comment]
@antonfirsov It would be useful if you could confirm I'm not doing anything too critical/stupid fiddling with the `MemoryDiagnostics` class.

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs
Comment (br3aker) [inline src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs:210]: There's no need to do a try-catch here I believe as it is handled by the caller here:

https://github.com/SixLabors/ImageSharp/blob/136cc143bba2fca255328751dd4580e4f8d413c3/src/ImageSharp/Formats/Jpeg/JpegDecoder.cs#L19-L26

Comment (tocsoft) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs]: yeah looks like you are right... I think I must have added it on first pass trying to find the leaky code.

Comment (br3aker) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs]: It's very tricky because jpeg is the only decoder which has a disposable internal `DecoderCore` implementation :)

Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs]: Should we make it so it’s doesn’t implement IDisposable?

Comment (br3aker) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs]: Hmm... Considering it's only used for `Frame` field which is a class field for the sake of not passing it to a bunch of methods we can make it non-disposable for consistency.

If you do so, try-catch must be replaced by try-finally or `Frame` would leak in a no-exception scenario. Also don't forget about `Identify` method - it also creates a `Frame` field so it should have a proper try-finally too.

**Review by JimBobSquarePants — APPROVED** [review]
Ah this is fantastic! Exactly what we need. 😍

File: src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Jpeg/JpegDecoderCore.cs]: Thanks. Ace, that’s what I thought. That can be a separate PR though.

**Review by br3aker — APPROVED** [review]
`ValidateDisposedMemoryAllocationsAttribute` is a blessing, also cool that current coverage reports no leaks :sunglasses:

**Comment by tocsoft** [comment]
> `ValidateDisposedMemoryAllocationsAttribute` is a blessing, also cool that current coverage reports no leaks 😎

no leaks (anymore) in the few tests I added it to anyway :D 

There will be a need to follow up adding this to a lot more of the test... I didn't want to blow out this PR fixing leaks hiding the framework changes.

Despite this being approved I think its worth leave this one open a bit longer before merging to give anyone with more knowledge on how my changed to the `MemoryDiagnostics` class might impact perf.

File: src/ImageSharp/Diagnostics/MemoryDiagnostics.cs
Comment (antonfirsov) [inline src/ImageSharp/Diagnostics/MemoryDiagnostics.cs:81]: `TotalUndisposedAllocationCount` was intended to report global allocation count. If I get the docs of [`AsyncLocal<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=net-6.0) right, the PR changes the semantics, so it becomes specific to an asynchronous execution flow, and will reset outside of the flow. I don't think this public behavior change is beneficial for the users of `MemoryDiagnostics.TotalUndisposedAllocationCount `. We need a separate API for `MemoryAllocatorValidator`.

File: src/ImageSharp/Formats/Tiff/Compression/Decompressors/JpegTiffCompression.cs
Comment (antonfirsov) [inline src/ImageSharp/Formats/Tiff/Compression/Decompressors/JpegTiffCompression.cs:69]: Do we also have separate PR-s for the Jpeg+Tiff+WebP+Bmp fixes or just for PNG?

File: tests/ImageSharp.Tests/MemoryAllocatorValidator.cs
Comment (antonfirsov) [inline tests/ImageSharp.Tests/MemoryAllocatorValidator.cs:13]:
```suggestion
        public static TestMemoryAllocatorDisposable MonitorAllocations(int max = 0)
```
No point for `TestMemoryAllocatorDisposable` being a struct, if we box it into `IDisposable` in the end.

File: tests/ImageSharp.Tests/MemoryAllocatorValidator.cs
Comment (antonfirsov) [inline tests/ImageSharp.Tests/MemoryAllocatorValidator.cs:19]: Is there a use case for `max > 0`? Do we want to allow any test with monitoring enabled to leak *some* memory?

**Review by antonfirsov — CHANGES_REQUESTED** [review]
We need to separate the `AsyncLocal` test-scoped validation from the global `MemoryDiagnostics.TotalUndisposedAllocationCount` counter.

**Comment by antonfirsov** [comment]
What I would do is to expose (for now internal) `BufferAllocated` / `BufferDestroyed` events and do something like:

```C#
internal static void IncrementTotalUndisposedAllocationCount()
{
    Interlocked.Increment(ref totalUndisposedAllocationCount);
    bufferAllocatedBackingDelegate?.Invoke();
}
```

Then move the `AsyncLocal` utility to test code and control it by subscribing to the events above.

File: src/ImageSharp/Diagnostics/MemoryDiagnostics.cs
Comment (tocsoft) [reply src/ImageSharp/Diagnostics/MemoryDiagnostics.cs:81]: It way its setup it will only report the local context when its been overwritten in tests (internal api)   by default the backing fields are backs by the `MemoryDiagnostics.Default` static instance which will be used in production usage outside of testing.

**Comment by tocsoft** [comment]
> What I would do is to expose (for now internal) `BufferAllocated` / `BufferDestroyed` events and do something like:
> 
> ```cs
> internal static void IncrementTotalUndisposedAllocationCount()
> {
>     Interlocked.Increment(ref totalUndisposedAllocationCount);
>     bufferAllocatedBackingDelegate?.Invoke();
> }
> ```
> 
> Then move the `AsyncLocal` utility to test code and control it by subscribing to the events above.

I think I like that better its less impactful on the prod code.... i'll have a play and see how feels.

File: src/ImageSharp/Diagnostics/MemoryDiagnostics.cs
Comment (antonfirsov) [reply src/ImageSharp/Diagnostics/MemoryDiagnostics.cs:81]: I see now, I missed the purpose of the branching. Though probably still better to keep the internal API as minimalistic and non-intrusive as possible.

File: tests/ImageSharp.Tests/MemoryAllocatorValidator.cs
Comment (antonfirsov) [inline tests/ImageSharp.Tests/MemoryAllocatorValidator.cs:60]: `expectedAllocationCount` seems to be always 0.

File: tests/ImageSharp.Tests/MemoryAllocatorValidator.cs
Comment (tocsoft) [reply tests/ImageSharp.Tests/MemoryAllocatorValidator.cs:56]: I was pre-empting being useful for tests that want to track/verify total number of allocated buffers not just retained buffers... but your right currently not being used.

File: tests/ImageSharp.Tests/MemoryAllocatorValidator.cs
Comment (tocsoft) [reply tests/ImageSharp.Tests/MemoryAllocatorValidator.cs:60]: In our current tests your are right, but I was envisioning use cases where we might want to want to validate before all buffers are returned and this is to allow for testing a specific/expected number of buffers are currently allocated still.

File: tests/ImageSharp.Tests/MemoryAllocatorValidator.cs
Comment (antonfirsov) [reply tests/ImageSharp.Tests/MemoryAllocatorValidator.cs:56]: I'm generally reluctant to add unused/untested features (= stuff to maintain) even to test utility code. When actual need appears, we can easily add them.

But it's really a nit, so I'm not going to block the PR on this.

**Review by antonfirsov — APPROVED** [review]

**Pull request merged**
